wordpress-mobile/WordPress-Android
# Issue/9847 use workmanager to postpone uploadstarter run

## Manual testing of the periodic task

Manual testing is a bit tricky, I get this message sometimes, and I'm not sure why:

`D WM-WorkerWrapper: Delaying execution for org.wordpress.android.util.AutoUploadWorker because it is being executed before schedule.`

- Start the app.
- Close the app.
- Make sure your device is not on data roaming.
- Go to your device settings, update the time (+7hours).
- Wait for WM to run the task.
- Check/grep logcat, you should get:

```
05-22 16:30:19.525 12364 12458 D WM-WorkerWrapper: org.wordpress.android.util.AutoUploadWorker returned a Success {mOutputData=androidx.work.Data@0} result.
05-22 16:30:19.535 12364 12458 I WM-WorkerWrapper: Worker result SUCCESS for Work [ id=75d6cbf2-3a46-4b2f-a1ad-db278b87d374, tags={ org.wordpress.android.util.AutoUploadWorker } ]
```

Tip: you can watch WorkManager logs by grepping "WM-" in logcat: `adb logcat | grep -i "WM-"`

## Manual testing of the onetime task

- Start the app in airplane mode
- Create a new post, write a title, press back
- Close the app.
- Turn off airplane mode.
- Make sure your device is not on data roaming.
- Wait for WM to run the task (should be instant).
- Check/grep logcat, you should get:

```
05-23 22:46:09.261 20329 20367 D WM-NetworkStateTracker: Network capabilities changed: [ Transports: CELLULAR Capabilities: MMS&SUPL&DUN&FOTA&IMS&CBS&IA&INTERNET&NOT_RESTRICTED&TRUSTED&NOT_VPN&VALIDATED&NOT_ROAMING&FOREGROUND&NOT_CONGESTED&NOT_SUSPENDED Unwanted:  LinkUpBandwidth>=51200Kbps LinkDnBandwidth>=102400Kbps Specifier: <1>]
05-23 22:46:09.268 20329 20365 D WM-Processor: Processor: processing a090b5c0-4767-402a-8752-eeaca683b61e
05-23 22:46:09.784 20329 20329 D WM-WorkerWrapper: Starting work for org.wordpress.android.util.AutoUploadWorker
05-23 22:46:09.794 20329 20365 D WM-WorkerWrapper: org.wordpress.android.util.AutoUploadWorker returned a Success {mOutputData=androidx.work.Data@0} result.
05-23 22:46:09.795 20329 20365 I WM-WorkerWrapper: Worker result SUCCESS for Work [ id=a090b5c0-4767-402a-8752-eeaca683b61e, tags={ org.wordpress.android.util.AutoUploadWorker } ]
```

## Instrumented tests

```sh
$ ./gradlew WordPress:cAT -Pandroid.testInstrumentationRunnerArguments.class=org.wordpress.android.util.UploadWorkRequestTest
```



Notes:

- I set up an 8 hours periodic with a 6 hours flex time. The main purpose of having a flex time is to delay the initial call (we're already calling `queueUploadFromAllSites` when the app starts). `queueUploadFromAllSites` function won't do anything if there is nothing to do, so I think it's fine to call it maximum 4 times a day.
- I don't like having the `MockedUploadStarterModule` and `MockedUploadAppComponent` setup, but since we can't use constructor injection in the UploadWorker (we're not in control of this object instantiation), we have do inject mocks via a specific module.
- I couldn't mock `LocalDraftUploadStarter` directly via Mockito, so I wrote a fake class: FakeLocalDraftUploadStarter, I'd prefer to use Mockito directly and use `whenever/verify`, but I couldn't find a solution to bypass a null CoroutineContext issue.
- I added WorkManager debug logging in `WordPressDebug` in b9db468 
- <s> ⚠️ a246fb7 ⚠️ - I'm pretty sure this line was a mistake. I'll investigate more about this and about the usage of `mOriginalPost` in `EditPostActivity`. I feel that it was used by different devs for different purposes (check for recent changes, and check for session change). I might revert this commit and open a new ticket instead.</s> Reverted
- I got weird errors with `mockito-core` and I had to use `mockito-android` instead, I also removed dependencies to `objenesis` and `byte-buddy` since there are dependencies of `mockito-android:`
```
+--- org.mockito:mockito-android:2.27.0
     +--- org.mockito:mockito-core:2.27.0
     |    +--- net.bytebuddy:byte-buddy:1.9.10
     |    +--- net.bytebuddy:byte-buddy-agent:1.9.10
     |    \--- org.objenesis:objenesis:2.6
     \--- net.bytebuddy:byte-buddy-android:1.9.10
          +--- net.bytebuddy:byte-buddy:1.9.10
          \--- com.jakewharton.android.repackaged:dalvik-dx:1
               \--- com.jakewharton.android.repackaged:libcore-dex:2
```
